### DISCUSSION TITLE
Provide some means of optionally using OpenSSL as backing cryptographic lib for FIPS applications

### DISCUSSION THREAD

**Discussion by josh-hemphill** [smallstep/certificates#712] [Ideas] [answered]
I'm not sure if FIPS certification has been addressed already, but looking at the available Go cryptographic libraries, it looks like there's OpenSSL wrappers that are api compatible with the native one. Providing some ability to swap the backing cryptographic code to OpenSSL could provide a pretty simple way for people who require FIPS certified libraries to still be able to smallstep and applications using the libs.

**Comment by maraino** [comment, marked as answer]
There's a couple of solutions for this:
1. Compile `step-ca` with Go's dev.boringcrypto branch, see [README.boringcrypto.md](https://github.com/golang/go/blob/dev.boringcrypto/README.boringcrypto.md), this branch is generally up-to-date with latest versions of Go and replaces crypto primitives with the ones using BoringSSL that is FIPS compliant, although officially they say this:
```
To be clear, we are not making any statements or representations about the suitability of this code in relation to 
the FIPS 140-2 standard. Interested users will have to evaluate for themselves whether the code is useful for their
own purposes.
```
For example, Rancher uses this image https://hub.docker.com/r/goboring/golang to build the "RKE Government" version of RKE2.

2. Redhat provides a FIPS version of Go that uses OpenSSL primitives, see the [announcemt](https://developers.redhat.com/blog/2019/06/24/go-and-fips-140-2-on-red-hat-enterprise-linux/). RedHat RPM packages are generally a few versions behind, but you can find more updated versions in their repositories. 

There's a [great article](https://gokulchandrapr.medium.com/go-crypto-and-kubernetes-fips-140-2-fedramp-compliance-66d852ccccd2) explaining this in more detail

**Reply by mmalone** [reply]
FYI, we plan to make fully-supported FIPS-compliant builds of `step-ca` available as part of our commercial offering. Shoot me a note (mike at smallstep) if that sounds interesting. FIPS compliance is niche enough that, at the moment, we can't justify the work to maintain a separate FIPS branch in open source.

**Reply by josh-hemphill** [reply]
> FYI, we plan to make fully-supported FIPS-compliant builds of `step-ca` available as part of our commercial offering. Shoot me a note (mike at smallstep) if that sounds interesting. FIPS compliance is niche enough that, at the moment, we can't justify the work to maintain a separate FIPS branch in open source.

I'm actually looking into this because I'm hoping to be able to use Caddyserver in FIPS environments, and Caddyserver uses the smallstep/certificates libs. It sounds like building using the RH FIPS fork of Go is still the solution I was looking for. As long as everything here doesn't break when I build with FIPS, then Caddy should be fine as well.

**Reply by mmalone** [reply]
Oof. Yea, that makes sense. I'd hate for us to be the blocker for FIPS builds of Caddy. We're super busy right now, but if you run into any specific issues as you proceed don't hesitate to reach out and we'll do what we can to help.

I'm also very curious to learn how easy the RH FIPS build process is / what (if anything) needs to change. Any notes you're willing to share would be greatly appreciated!

**Reply by josh-hemphill** [reply]
Will do. Thanks! :)

**Reply by maraino** [reply]
@josh-hemphill I've just compiled `step-ca` with `goboring/golang:1.16.7b7`, the latest image in docker hub, and `step-ca` compiles and runs fine. Do you know what Go versions are supported in RH FIPS?

**Reply by josh-hemphill** [reply]
Sorry, I don't have access to a licensed RHEL server at the moment to be able to find out.
If it's important, I could try asking some people who do.

**Reply by maraino** [reply]
Depending on the Go version that they support I can probably tell you if it will compile or not.

**Reply by mmalone** [reply]
FYI, you can get a RHEL-licensed server from AWS marketplace and pay by the hour. The process is pretty annoying (it's not as simple as launching a normal AMI) but I clicked around enough to get one spun up a couple weeks ago and it wasn't _too_ terrible. I think GCP / Azure have similar products.

https://aws.amazon.com/marketplace/pp/prodview-ubouaa3jksbmc
